receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+	defer jm.Unlock()
+	// once the structure is locked, confirm that gc() is still necessary
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		for sig, tsm := range jm.jobsMap {
+			tsm.RLock()
+			tsmNotMarked := !tsm.mark
+			tsm.RUnlock()
+			if tsmNotMarked {
+				delete(jm.jobsMap, sig)
+			} else {
+				tsm.gc()
+			}
+		}
+		jm.lastGC = time.Now()
+	}
+}
+
+func (jm *JobsMapPdata) maybeGC() {
+	// speculatively check if gc() is necessary, recheck once the structure is locked
+	jm.RLock()
+	defer jm.RUnlock()
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		go jm.gc()
+	}
+}
+
+func (jm *JobsMapPdata) get(job, instance string) *timeseriesMapPdata {
+	sig := job + ":" + instance
+	jm.RLock()
+	tsm, ok := jm.jobsMap[sig]
+	jm.RUnlock()
+	defer jm.maybeGC()
+	if ok {
+		return tsm
+	}
+	jm.Lock()
+	defer jm.Unlock()
+	tsm2, ok2 := jm.jobsMap[sig]
+	if ok2 {
+		return tsm2
+	}
+	tsm2 = newTimeseriesMapPdata()
+	jm.jobsMap[sig] = tsm2
+	return tsm2
+}
+
+// MetricsAdjusterPdata takes a map from a metric instance to the initial point in the metrics instance
+// and provides AdjustMetrics, which takes a sequence of metrics and adjust their start times based on
+// the initial points.
+type MetricsAdjusterPdata struct {
+	tsm    *timeseriesMapPdata
+	logger *zap.Logger
+}
+
+// NewMetricsAdjuster is a constructor for MetricsAdjuster.
+func NewMetricsAdjusterPdata(tsm *timeseriesMapPdata, logger *zap.Logger) *MetricsAdjusterPdata {
+	return &MetricsAdjusterPdata{
+		tsm:    tsm,
+		logger: logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their start times based on the initial and
+// previous points in the timeseriesMap.
+// Returns the total number of timeseries that had reset start times.
+func (ma *MetricsAdjusterPdata) AdjustMetrics(metricL *pdata.MetricSlice) int {
+	resets := 0
+	ma.tsm.Lock()
+	defer ma.tsm.Unlock()
+	for i := 0; i < metricL.Len(); i++ {
+		metric := metricL.At(i)
+		resets += ma.adjustMetric(&metric)
+	}
+	return resets
+}
+
+// Returns the number of timeseries with reset start times.
+func (ma *MetricsAdjusterPdata) adjustMetric(metric *pdata.Metric) int {
+	switch metric.DataType() {
+	case pdata.MetricDataTypeGauge:
+		// gauges don't need to be adjusted so no additional processing is necessary
+		return 0
+	default:
+		return ma.adjustMetricPoints(metric)
+	}
+}
+
+// Returns  the number of timeseries that had reset start times.
+func (ma *MetricsAdjusterPdata) adjustMetricPoints(metric *pdata.Metric) int {
+	switch dataType := metric.DataType(); dataType {
+	case pdata.MetricDataTypeGauge:
+		return ma.adjustMetricGauge(metric)
+
+	case pdata.MetricDataTypeHistogram:
+		return ma.adjustMetricHistogram(metric)
+
+	case pdata.MetricDataTypeSummary:
+		return ma.adjustMetricSummary(metric)
+
+	case pdata.MetricDataTypeSum:
+		return ma.adjustMetricSum(metric)
+
+	default:
+		// this shouldn't happen
+		ma.logger.Info("Adjust - skipping unexpected point", zap.String("type", dataType.String()))
```

Comment:
Debug level probably

Reply:
This is currently info in the OC metrics adjuster.  If anything, since this is a situation that should never occur I might want to make this a warning.